core/libp2p/host_builder.cpp

```diff
@@ -2,11 +2,173 @@
+
+  HostBuilder::HostBuilder(boost::asio::io_context &io_context)
+      : io_context_{io_context} {}
+
+  HostBuilder::HostBuilder(boost::asio::io_context &io_context, Config config)
+      : config_{std::move(config)}, io_context_{io_context} {}
+
+  HostBuilder &HostBuilder::setKeypair(const crypto::KeyPair &kp) {
+    config_.peer_key = kp;
+    return *this;
+  }
+
+  HostBuilder &HostBuilder::setCSPRNG(detail::sptr<crypto::random::CSPRNG> r) {
+    config_.c_prng = std::move(r);
+    return *this;
+  }
+
+  HostBuilder &HostBuilder::setPRNG(
+      detail::sptr<crypto::random::RandomGenerator> r) {
+    config_.prng = std::move(r);
+    return *this;
+  }
+
+  HostBuilder &HostBuilder::setRoutingAdaptor(
+      detail::sptr<routing::RoutingAdaptor> r) {
+    config_.routing = std::move(r);
+    return *this;
+  }
+
+  HostBuilder &HostBuilder::setDiscoveryAdaptor(
+      detail::sptr<discovery::DiscoveryAdaptor> d) {
+    config_.discovery = std::move(d);
+    return *this;
+  }
+
+  HostBuilder &HostBuilder::setPeerRepository(
+      detail::sptr<peer::PeerRepository> p) {
+    config_.peer_repository = std::move(p);
+    return *this;
+  }
+
+  HostBuilder &HostBuilder::addTransport(
+      detail::sptr<transport::Transport> tr) {
+    config_.transports.push_back(std::move(tr));
+    return *this;
+  }
+
+  HostBuilder &HostBuilder::addMuxerAdaptor(
+      detail::sptr<muxer::MuxerAdaptor> mux) {
+    config_.muxers.push_back(std::move(mux));
+    return *this;
+  }
+
+  HostBuilder &HostBuilder::addDHTAdaptor(detail::sptr<dht::DHTAdaptor> d) {
+    config_.dhts.push_back(std::move(d));
+    return *this;
+  }
+
+  HostBuilder &HostBuilder::addSecurityAdaptor(
+      detail::sptr<security::SecurityAdaptor> s) {
+    config_.securities.push_back(std::move(s));
+    return *this;
+  }
+
+  HostBuilder &HostBuilder::addListenMultiaddr(
+      const multi::Multiaddress &address) {
+    config_.listen_addresses.push_back(address);
+    return *this;
+  }
+
+  HostBuilder &HostBuilder::addListenMultiaddr(std::string_view address) {
+    auto addr_res = Multiaddress::create(address);
+    if (!addr_res) {
+      throw std::invalid_argument("cannot add listen multiaddress: "
+                                  + addr_res.error().message());
+    }
+    config_.listen_addresses.push_back(std::move(addr_res.value()));
+    return *this;
+  }
+
+  Host HostBuilder::build() {
+    if (!config_.c_prng) {
+      config_.c_prng = std::make_shared<crypto::random::BoostRandomGenerator>();
+    }
+
+    if (!config_.prng) {
+      config_.prng = std::make_shared<crypto::random::StdRandomGenerator>();
+    }
+
+    if (!keypairIsWellFormed(config_.peer_key)) {
+      crypto::KeyGeneratorImpl key_generator{*config_.c_prng};
+      auto keys_res = key_generator.generateKeys(crypto::Key::Type::RSA2048);
+      if (!keys_res) {
+        throw std::runtime_error("cannot generate a keypair");
+      }
+      config_.peer_key = std::move(keys_res.value());
```

Comment:
better create a helper (free `generate_keypair(&csprng);` func with this content)